presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartialColumnAdapter.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.block.BlockBuilder;
+import io.prestosql.spi.block.ColumnarRow;
+import io.prestosql.spi.block.LazyBlock;
+import io.prestosql.spi.block.LazyBlockLoader;
+import io.prestosql.spi.type.Type;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.prestosql.plugin.hive.HiveColumnHandle.ColumnType.REGULAR;
+import static io.prestosql.plugin.hive.HivePartialColumnAdapter.PartialColumnLoadingStrategy.READ_FULL_COLUMNS;
+import static io.prestosql.spi.block.ColumnarRow.toColumnarRow;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class HivePartialColumnAdapter
+{
+    public enum PartialColumnLoadingStrategy
```

Comment:
(Based on our discussion the other day...)

From an abstraction perspective, I think it might be cleaner to approach it this way:
* The connector describes what dereference paths it wants from the reader upon instantiating it
* The instantiation returns a reader + a list of (possibly incomplete) paths it's able to satisfy
* The connector inserts an adaptation layer around the reader if necessary to satisfy anything the reader is not able to produce

For example, let's say the connector promised the engine it will produce `a.x`, `a.y` and `b`. When it instantiates the reader, it provides `[a.x, a.y, b]` as its preferences. Let's suppose the reader can only read full columns. In such a case, it would return `[a, b]` as the list of what it can actually provide. The connector then inserts and adapter from `a` -> `[a.x, a.y]`.

I think this might fit well with other kinds of adaptations we might need to do, such as for type coercions. The connector might ask the reader to return something as a BIGINT, but the reader would say "I only support INTEGER". Then the connector would insert an adapter to convert the types.

Reply:
@wagnermarkd, how does this fit within the work you've been doing around schema evolution?